### ISSUE TITLE
Incorrect CS8670 warning for possible dereference in initializer that actually doesn't happen

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43287]
**Version Used**: 

```
Branch master (10 Mar 2020)
Latest commit 441c154 by msftbot[bot]:
Merge pull request #42303 from CyrusNajmabadi/useCompoundParens

Support parenthesized expressions in use-compound-operator.
```

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable
class C
{
  C? f = null;

  static void Main()
  {
    var c = new C() {f = { f = { }  }};
    c.f = new C();
  }
}
```

**Expected Behavior**:
No warnings. No dereferencing happens in the object initializer and the code runs without exceptions.

**Actual Behavior**:
`warning CS8670: Object or collection initializer implicitly dereferences possibly null member 'C.f'.`
